In [3]:
import pandas as pd
import numpy as np
from random import sample 

In [102]:
df = pd.read_excel('D:\Data Science Bootcamp\Day 5\dataset\data loan.xlsx')

In [103]:
## import copy
## df_edit = copy.deepcopy(df)
## df_edit['NPF'] = df_edit.NPF.map({'Ya':1,
##                                   'Tidak':0})

In [104]:
ln_income = np.log(df.Income)
ln_income.head()
df['ln_income'] = ln_income

In [105]:
# Random sampling - Random n rows
df_elements = df.sample(n=4)
print(df_elements)

       Kontrak   Tipe_Kendaraan   Pekerjaan      Area  Tenor  \
18006    18007          MINIBUS  Wiraswasta      Jawa     36   
22674    22675             LCGC  Wiraswasta    Others     60   
26638    26639  TRUCK & PICK UP  Wiraswasta  Sumatera     48   
18181    18182             LCGC  Wiraswasta  Sumatera     60   

      Bucket_Collectability    NPF    Total_AR     LTV     OTR  ...      AF  \
18006           0 - 30 hari  Tidak  210.758771   74.52  253.95  ...  189.24   
22674           0 - 30 hari  Tidak  272.159996  102.39  175.15  ...  179.34   
26638            > 150 hari     Ya  367.169018   76.91  327.00  ...  251.49   
18181           0 - 30 hari  Tidak  212.280002   97.47  133.80  ...  130.42   

        Tujuan  Pendidikan  Tanggungan   Status     Gender  Paket Inflasi  \
18006  Private      Others         Ada   OTHERS  Perempuan  Tidak    8.61   
22674  Private      Others         Ada  MENIKAH  Perempuan     Ya    3.72   
26638   Others      Others         Ada   OTHERS  Lak

In [106]:
# Random sampling - Random n% rows
df_80pct = df.sample(frac=0.8)
print(df_80pct)

       Kontrak   Tipe_Kendaraan   Pekerjaan      Area  Tenor  \
13596    13597          MINIBUS    Karyawan      Jawa     48   
21164    21165          MINIBUS  Wiraswasta    Others     48   
20218    20219          MINIBUS  Wiraswasta    Others     36   
6362      6363          MINIBUS    Karyawan  Sumatera     24   
19909    19910           OTHERS    Karyawan      Jawa     48   
...        ...              ...         ...       ...    ...   
11219    11220          MINIBUS    Karyawan  Sumatera     48   
24106    24107          MINIBUS  Wiraswasta      Jawa     60   
25384    25385          MINIBUS    Karyawan      Jawa     48   
12561    12562          MINIBUS    Karyawan      Jawa     48   
27100    27101  TRUCK & PICK UP  Wiraswasta    Others     48   

      Bucket_Collectability    NPF    Total_AR    LTV     OTR  ...      AF  \
13596           0 - 30 hari  Tidak  175.488001  83.40  165.16  ...  137.74   
21164       >120 - 150 hari     Ya  246.288004  86.02  220.55  ...  189.73 

In [107]:
df_80pct.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 22634 entries, 13596 to 27100
Data columns (total 24 columns):
Kontrak                  22634 non-null int64
Tipe_Kendaraan           22634 non-null object
Pekerjaan                22634 non-null object
Area                     22634 non-null object
Tenor                    22634 non-null int64
Bucket_Collectability    22634 non-null object
NPF                      22634 non-null object
Total_AR                 22634 non-null float64
LTV                      22634 non-null float64
OTR                      22634 non-null float64
Age                      22634 non-null int64
Interest_Rate            22634 non-null float64
Income                   22634 non-null int64
DP                       22634 non-null float64
AF                       22634 non-null float64
Tujuan                   22634 non-null object
Pendidikan               22634 non-null object
Tanggungan               22634 non-null object
Status                   22634 non-null

In [108]:
import statsmodels.formula.api as smf

In [112]:
ols_df_80pct = smf.ols('Total_AR ~ Income', data = df_80pct).fit()
print(ols_df_80pct.summary())

                            OLS Regression Results                            
Dep. Variable:               Total_AR   R-squared:                       0.003
Model:                            OLS   Adj. R-squared:                  0.003
Method:                 Least Squares   F-statistic:                     69.89
Date:                Fri, 25 Oct 2019   Prob (F-statistic):           6.62e-17
Time:                        17:46:20   Log-Likelihood:            -1.3347e+05
No. Observations:               22634   AIC:                         2.670e+05
Df Residuals:                   22632   BIC:                         2.670e+05
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept    213.8293      0.596    359.055      0.0

In [113]:
ols_df_80pct = smf.ols('Total_AR ~ ln_income', data = df_80pct).fit()
print(ols_df_80pct.summary())

                            OLS Regression Results                            
Dep. Variable:               Total_AR   R-squared:                       0.040
Model:                            OLS   Adj. R-squared:                  0.040
Method:                 Least Squares   F-statistic:                     944.2
Date:                Fri, 25 Oct 2019   Prob (F-statistic):          3.53e-203
Time:                        17:46:23   Log-Likelihood:            -1.3305e+05
No. Observations:               22634   AIC:                         2.661e+05
Df Residuals:                   22632   BIC:                         2.661e+05
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept    149.8307      2.189     68.441      0.0

# I tried to compare the Income variable and ln_income variable to see if the log(income) affects the model or not. It turns out the logarithmic income does not affect so much, but the model is little better.
### The R-Squared for Income is 0.003
### The R-Squared for ln_income is 0.04 (a little bit stronger)

In [114]:
ols_df_80pct = smf.ols('Total_AR ~ OTR', data = df_80pct).fit()
print(ols_df_80pct.summary())

                            OLS Regression Results                            
Dep. Variable:               Total_AR   R-squared:                       0.538
Model:                            OLS   Adj. R-squared:                  0.538
Method:                 Least Squares   F-statistic:                 2.633e+04
Date:                Fri, 25 Oct 2019   Prob (F-statistic):               0.00
Time:                        17:49:21   Log-Likelihood:            -1.2478e+05
No. Observations:               22634   AIC:                         2.496e+05
Df Residuals:                   22632   BIC:                         2.496e+05
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     49.0633      1.096     44.759      0.0

# The OTR variable makes the model way stronger than the ln_income variable itself, with the R-Squared of 0.538.


In [115]:
ols_df_80pct = smf.ols('Total_AR ~ AF', data = df_80pct).fit()
print(ols_df_80pct.summary())

                            OLS Regression Results                            
Dep. Variable:               Total_AR   R-squared:                       0.913
Model:                            OLS   Adj. R-squared:                  0.913
Method:                 Least Squares   F-statistic:                 2.362e+05
Date:                Fri, 25 Oct 2019   Prob (F-statistic):               0.00
Time:                        17:51:29   Log-Likelihood:            -1.0593e+05
No. Observations:               22634   AIC:                         2.119e+05
Df Residuals:                   22632   BIC:                         2.119e+05
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     -0.7251      0.476     -1.523      0.1

# The AF variable makes the model way stronger than the OTR variable itself, with the R-Squared off 0.913.

If we combine all three variables in one model, the model is stronger a little bit with R-Squared of 0.937

In [116]:
ols_df_80pct = smf.ols('Total_AR ~ ln_income + OTR + AF', data = df_80pct).fit()
print(ols_df_80pct.summary())

                            OLS Regression Results                            
Dep. Variable:               Total_AR   R-squared:                       0.937
Model:                            OLS   Adj. R-squared:                  0.937
Method:                 Least Squares   F-statistic:                 1.118e+05
Date:                Fri, 25 Oct 2019   Prob (F-statistic):               0.00
Time:                        17:54:34   Log-Likelihood:            -1.0226e+05
No. Observations:               22634   AIC:                         2.045e+05
Df Residuals:                   22630   BIC:                         2.046e+05
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     20.1117      0.609     33.044      0.0

From the quotient above, we get Total_AR = 20.1117 - 5.5420ln_income - 0.2956 OTR + 1.6493AF
- If the AF goes up by IDR 1Mio, then the Total_AR will increase by IDR 1.65Mio.
- If the ln_income goes up by IDR 1Mio, then the Total_AR will decrease by IDR 5.5Mio.
- If the OTR goes up by IDR 1Mio, then the Total_AR will decrease by IDR 295k.